### ISSUE TITLE
MongoDB text search does not support Star Wars languages

### ISSUE THREAD

**Issue by r-marques** [bigchaindb/bigchaindb#1670]
When creating a text index with MongoDB the index is created with a default language (usually en).
MongoDB also allow us to override the language used to create the text index for a particular document by adding the field `language` inside the document being written. From the [documentation](https://docs.mongodb.com/manual/tutorial/specify-language-for-text-index/):

> If a collection contains documents or embedded documents that are in different languages, include a field named language in the documents or embedded documents and specify as its value the language for that document or embedded document.
MongoDB will use the specified language for that document or embedded document when building the text index

If you specify a language that is not supported by MongoDB the insert operation will fail. This is a problem for us because we don't know how BigchainDB clients will use the asset field.

For example I was testing with star wars data from [swapi.co](http://swapi.co/):
```json
{
  "average_height": "300",
  "language": "Huttese",
  "url": "http://swapi.co/api/species/5/",
  "eye_colors": "yellow, red",
  "name": "Hutt",
  "hair_colors": "n/a",
  "people": [
    "http://swapi.co/api/people/16/"
  ],
  "edited": "2014-12-20T21:36:42.146000Z",
  "homeworld": "http://swapi.co/api/planets/24/",
  "classification": "gastropod",
  "films": [
    "http://swapi.co/api/films/3/",
    "http://swapi.co/api/films/1/"
  ],
  "designation": "sentient",
  "average_lifespan": "1000",
  "skin_colors": "green, brown, tan",
  "created": "2014-12-10T17:12:50.410000Z"
}
```

When trying to insert this asset MongoDB fails with `WriteError: language override unsupported: Huttese`.

We need to somehow either disable language override so that `language` fields in documents do not affect how the text index is built, or default to some language when there is an error.

**Comment by TimDaub** [comment]
> We need to somehow either disable language override so that language fields in documents do not affect how the text index is built, or default to some language when there is an error.

I think we should rather ask MongoDB to support Huttese :rofl:

**Comment by vrde** [comment]
> I think we should rather ask MongoDB to support Huttese

Agree with @TimDaub, you should open an issue on their Jira. How can we support the Interplanetary Database foundation if we don't support Huttese? This issue is top priority.

**Comment by TimDaub** [comment]
> How can we support the Interplanetary Database foundation if we don't support Huttese?

:+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1:

**Comment by ttmc** [comment]
Yesterday this issue caused problems for one of our key users. They had an asset which looked like so:
```json
   "asset":{
      "data":{
         "uuid":"75be7c49-315f-4ba8-9010-37efa00d5f30",
         "family":"Office",
         "vendor":"Microsoft",
         "edition":"Standard",
         "version":"2007",
         "language":"single"
      }
   }
```

MongoDB doesn't know of any language named `"single"`.

**Comment by ttmc** [comment]
I was thinking more about this issue and realized that MongoDB's support for language-specific text search is actually a _feature_. Maybe we could take advantage of that (in our asset text search endpoint), rather than trying to convince MongoDB to give us a way to turn it off?

If someone submits a CREATE transaction with an asset containing the `"language"` field/key that's _not_ recognized by MongoDB, then maybe we could return a helpful error message telling them why their transaction is being rejected as invalid, and suggest that they use a different field/key name, such as `"lang"`.

Quietly converting an unrecognized `"language"` to `"English"` would break many legit user assets, such as the one above (`"language":"single"`). Let's not do that.

**Comment by kansi** [comment]
I agree with @ttmc that a default fallback language might lead to undesired concequences. Since there is a finite [list of languages](https://docs.mongodb.com/manual/reference/text-search-languages/) that mongodb supports, we could validate `language` parameter in `data` and return a `400` in case the language is not supported?

**Comment by ttmc** [comment]
Yes, I agree. And if they _do_ give an invalid value for language, be sure to respond with an informative error message, e.g. MongoDB doesn't support text search for the language `"klingon"`. If you don't care about text search and want to record something has having that language, then we suggest renaming the key/field to something other than `"language"` such as `"lang"`.

It would be unfortunate if they already stored a bunch of assets (transactions) using the `"language`" key/field. Maybe we could return an informational message (along with `200`) when they use `"language"`, to warn them about potential problems.

@sbellem or @vrde  What do you think?

**Comment by kansi** [comment]
@ttmc @sbellem @vrde I think there are two ways we could solve this while maintaing backward compatibility,
(1) update the `version` key
(2) Only update the json schema for transaction

Furthermore, for users using MongoDB as backend this constrain should not affect them as they cannot have any asset in db with key `langugage` whose value is something other than what is supported by MongoDB.
But for users using RethinkDB, it might cause an issue because it doesn't give any special treatment to `language`.

So, it would make sense to apply the schema level validation fix in 2.0 release of BDB. For BDB 1.x we can handle this at the web api level.

**Comment by ttmc** [comment]
We can't update the allowed values in the transaction schema (including the allowed values of `version`) without breaking backwards compatibility, which is something we said we wouldn't do in the 1.x series. I like the idea of handling it at the web API level for now.

**Comment by kansi** [comment]
closed #1806

**Comment by mfaheemakhtar** [comment]
For future time-travelers, you can use `language_override` property in the text index to change the `language` key.

```
Schema.index({ name: 'text' }, { language_override: 'anotherKey' });
```

Ref: https://docs.mongodb.com/manual/tutorial/specify-language-for-text-index/#use-any-field-to-specify-the-language-for-a-document
